core/src/main/java/org/apache/brooklyn/core/mgmt/ha/OsgiArchiveInstaller.java

```diff
@@ -0,0 +1,429 @@
+    
+    final private OsgiManager osgiManager;
+    private ManagedBundle suppliedKnownBundleMetadata;
+    private InputStream zipIn;
+    
+    private boolean start = true;
+    private boolean loadCatalogBom = true;
+    private boolean force = false;
+    private boolean deferredStart = false;
+    
+    private File zipFile;
+    private Manifest discoveredManifest;
+    private VersionedName discoveredBomVersionedName;
+    OsgiBundleInstallationResult result;
+    
+    private ManagedBundle inferredMetadata;
+    private final boolean inputStreamSupplied;
+    
+    OsgiArchiveInstaller(OsgiManager osgiManager, ManagedBundle knownBundleMetadata, InputStream zipIn) {
+        this.osgiManager = osgiManager;
+        this.suppliedKnownBundleMetadata = knownBundleMetadata;
+        this.zipIn = zipIn;
+        inputStreamSupplied = zipIn!=null;
+    }
+
+    public void setStart(boolean start) {
+        this.start = start;
+    }
+    
+    public void setLoadCatalogBom(boolean loadCatalogBom) {
+        this.loadCatalogBom = loadCatalogBom;
+    }
+    
+    public void setForce(boolean force) {
+        this.force = force;
+    }
+
+    public void setDeferredStart(boolean deferredStart) {
+        this.deferredStart = deferredStart;
+    }    
+
+    private ManagementContextInternal mgmt() {
+        return (ManagementContextInternal) osgiManager.mgmt;
+    }
+    
+    private synchronized void init() {
+        if (result!=null) {
+            if (zipFile!=null || zipIn==null) return;
+            throw new IllegalStateException("This installer instance has already been used and the input stream discarded");
+        }
+        result = new OsgiBundleInstallationResult();
+        inferredMetadata = suppliedKnownBundleMetadata==null ? new BasicManagedBundle() : suppliedKnownBundleMetadata;
+    }
+    
+    private synchronized void makeLocalZipFileFromInputStreamOrUrl() {
+        if (zipIn==null) {
+            Maybe<Bundle> installedBundle = Maybe.absent();
+            if (suppliedKnownBundleMetadata!=null) {
+                // if no input stream, look for a URL and/or a matching bundle
+                if (!suppliedKnownBundleMetadata.isNameResolved()) {
+                    ManagedBundle mbFromUrl = osgiManager.getManagedBundleFromUrl(suppliedKnownBundleMetadata.getUrl());
+                    if (mbFromUrl!=null) {
+                        // user supplied just a URL (eg brooklyn.libraries), but we recognise it,
+                        // so don't try to reload it, just record the info we know about it to retrieve the bundle
+                        ((BasicManagedBundle)suppliedKnownBundleMetadata).setSymbolicName(mbFromUrl.getSymbolicName());
+                        ((BasicManagedBundle)suppliedKnownBundleMetadata).setVersion(mbFromUrl.getVersion());
+                    }
+                }
+                if (installedBundle.isAbsent() && suppliedKnownBundleMetadata.getOsgiUniqueUrl()!=null) {
```

Comment:
there's an `isAbsent` check in each `if` test and the result of each body might be an absent so `else wouldn't be sufficient